File: Sources/DomoBrokerCore/MCPSession.swift
import Foundation
import DomoProtocol
import DomoTransport

/// One authenticated agent connection speaking MCP (JSON-RPC 2.0, one message
/// per line). Translates tool calls into signed intents routed to devices.
final class MCPSession {
    private unowned let broker: Broker
    private let conn: Connection
    private let agentToken: String
    private let sessionId = UUID().uuidString
    private let queue = DispatchQueue(label: "domo.mcp.session", attributes: .concurrent)

    init(broker: Broker, connection: Connection, agentToken: String) {
        self.broker = broker
        self.conn = connection
        self.agentToken = agentToken
    }

    private var agent: AgentRecord? { broker.store.agent(token: agentToken) }

    func handleLine(_ line: Data) {
        guard let message = try? JSONValue.parse(line) else { return }
        let id = message["id"]
        guard let method = message["method"].str else { return }
        if id.isNull { return } // notifications (e.g. notifications/initialized)
        queue.async { [weak self] in
            guard let self else { return }
            let result = self.dispatch(method: method, params: message["params"])
            switch result {
            case .success(let value):
                self.send(["jsonrpc": "2.0", "id": id, "result": value])
            case .failure(let error):
                self.send(["jsonrpc": "2.0", "id": id,
                           "error": ["code": .number(-32000), "message": .string(error.message)]])
            }
        }
    }

    private func send(_ message: JSONValue) {
        conn.sendLine(message.encoded())
    }

    private func dispatch(method: String, params: JSONValue) -> Result<JSONValue, RPCError> {
        switch method {
        case "initialize":
            let requested = params["protocolVersion"].str ?? "2024-11-05"
            return .success([
                "protocolVersion": .string(requested),
                "capabilities": ["tools": [:]],
                "serverInfo": ["name": "domo-broker", "version": "0.1.0"],
            ])
        case "ping":
            return .success([:])
        case "tools/list":
            return .success(["tools": MCPTools.definitions])
        case "tools/call":
            guard let name = params["name"].str else {
                return .failure(RPCError("missing tool name"))
            }
            return callTool(name: name, args: params["arguments"])
        default:
            return .failure(RPCError("method not supported: \(method)"))
        }
    }

    // MARK: - Tool dispatch

    private func callTool(name: String, args: JSONValue) -> Result<JSONValue, RPCError> {
        do {
            let result = try runTool(name: name, args: args)
            return .success(["content": [["type": "text", "text": .string(result.jsonString())]],
                             "isError": false])
        } catch let error as ToolError {
            return .success(["content": [["type": "text", "text": .string(error.message)]],
                             "isError": true])
        } catch {
            return .success(["content": [["type": "text", "text": .string("\(error)")]],
                             "isError": true])
        }
    }

    struct ToolError: Error {
        let message: String
        init(_ message: String) { self.message = message }
    }

    private func runTool(name: String, args: JSONValue) throws -> JSONValue {
        switch name {
        case "list_devices": return try listDevices()
        case "request_device_access": return try requestAccess(args)
        case "read_file": return try readFile(args)
        case "write_file": return try writeFile(args)
        case "run_command": return try runCommand(args)
        case "get_output": return try getOutput(args)
        case "list_device_tools": return try listDeviceTools(args)
        case "use_tool": return try useTool(args)
        default: throw ToolError("unknown tool: \(name)")
        }
    }

    private func requireAgent() throws -> AgentRecord {
        guard let agent else { throw ToolError("agent identity revoked") }
        return agent
    }

    private func requireGrantedLink(_ args: JSONValue) throws -> (AgentRecord, String, Broker.DeviceLink) {
        let agent = try requireAgent()
        guard let deviceId = args["device"].str else { throw ToolError("missing 'device'") }
        guard agent.grantedDevices.contains(deviceId) else {
            throw ToolError("no grant for device \(deviceId); call request_device_access first")
        }
        guard let link = broker.deviceLink(deviceId) else {
            throw ToolError("device \(deviceId) is offline")
        }
        return (agent, deviceId, link)
    }

    private func listDevices() throws -> JSONValue {
        let agent = try requireAgent()
        let online = broker.onlineDeviceIds()
        return ["devices": .array(broker.store.allDevices().map { device in
            [
                "id": .string(device.deviceId),
                "name": .string(device.name),
                "online": .bool(online.contains(device.deviceId)),
                "granted": .bool(agent.grantedDevices.contains(device.deviceId)),
            ]
        })]
    }

    private func requestAccess(_ args: JSONValue) throws -> JSONValue {
        let agent = try requireAgent()
        guard let deviceId = args["device"].str else { throw ToolError("missing 'device'") }
        guard let goals = args["goals"].str, !goals.isEmpty else {
            throw ToolError("missing 'goals' — state what you intend to do on this Mac")
        }
        guard let link = broker.deviceLink(deviceId) else {
            throw ToolError("device \(deviceId) is offline")
        }
        if agent.grantedDevices.contains(deviceId) {
            return ["status": "granted", "note": "already granted"]
        }
        // Approval can take as long as a human takes; generous timeout.
        let response = try link.rpc.call("access_request", [
            "agent": [
                "id": .string(agent.agentId),
                "display": .string(agent.display),
                "publicKey": .string(agent.publicKeyBase64),
            ],
            "goals": .string(goals),
        ], timeout: 600)
        let approved = response["approved"].boolValue ?? false
        if approved {
            broker.store.grantDevice(token: agentToken, deviceId: deviceId)
            broker.store.recordSessionGoals(token: agentToken, goals: goals)
        }
        return ["status": .string(approved ? "granted" : "denied")]
    }

    // MARK: - Intent construction

    private func makeIntent(agent: AgentRecord, deviceId: String, goal: String?,
                            request: String, capabilities: [Capability]) throws -> Intent {
        var intent = Intent(agentId: agent.agentId, agentDisplay: agent.display,
                            agentPublicKey: agent.publicKeyBase64, deviceId: deviceId,
                            goal: goal, planContext: agent.sessionGoals,
                            request: request, capabilities: capabilities,
                            sessionId: sessionId)
        try intent.sign(with: try agent.keyPair())
        return intent
    }

    private func sendIntent(_ intent: Intent, payload: JSONValue,
                            to link: Broker.DeviceLink, timeout: TimeInterval = 630) throws -> JSONValue {
        let response = try link.rpc.call("intent", [
            "intent": JSONValue.from(intent),
            "payload": payload,
        ], timeout: timeout)
        switch response["status"].str {
        case "denied": throw ToolError("the device owner denied this request")
        case "rejected": throw ToolError("device rejected intent: \(response["reason"].str ?? "unknown")")
        case "error": throw ToolError(response["error"].str ?? "device error")
        default: return response
        }
    }

    // MARK: - Tools

    private func readFile(_ args: JSONValue) throws -> JSONValue {
        let (agent, deviceId, link) = try requireGrantedLink(args)
        guard let path = args["path"].str else { throw ToolError("missing 'path'") }
        let intent = try makeIntent(agent: agent, deviceId: deviceId, goal: args["goal"].str,
                                    request: "read file: \(path)",
                                    capabilities: [Capability(kind: .fsRead, paths: [path])])
        let response = try sendIntent(intent, payload: .null, to: link)
        guard let base64 = response["content_base64"].str, let data = Data(base64Encoded: base64) else {
            throw ToolError("device returned no content")
        }
        if let text = String(data: data, encoding: .utf8) {
            return ["path": .string(path), "content": .string(text)]
        }
        return ["path": .string(path), "content_base64": .string(base64)]
    }

    private func writeFile(_ args: JSONValue) throws -> JSONValue {
        let (agent, deviceId, link) = try requireGrantedLink(args)
        guard let path = args["path"].str else { throw ToolError("missing 'path'") }
        guard let content = args["content"].str else { throw ToolError("missing 'content'") }
        let data = Data(content.utf8)
        let intent = try makeIntent(agent: agent, deviceId: deviceId, goal: args["goal"].str,
                                    request: "write file: \(path) (\(data.count) bytes)",
                                    capabilities: [Capability(kind: .fsWrite, paths: [path])])
        let response = try sendIntent(intent, payload: [
            "content_base64": .string(data.base64EncodedString()),
        ], to: link)
        return ["path": .string(path), "bytes": response["bytes"]]
    }

    private func runCommand(_ args: JSONValue) throws -> JSONValue {
        let (agent, deviceId, link) = try requireGrantedLink(args)
        guard let argvValues = args["argv"].arr, !argvValues.isEmpty else {
            throw ToolError("missing 'argv'")
        }
        let argv = argvValues.compactMap { $0.str }
        guard argv.count == argvValues.count else { throw ToolError("argv must be strings") }
        let readPaths = (args["read_paths"].arr ?? []).compactMap { $0.str }
        let writePaths = (args["write_paths"].arr ?? []).compactMap { $0.str }
        let network = args["network"].boolValue ?? false
        let waitMs = args["wait_ms"].int ?? 10000

        var capabilities: [Capability] = [
            Capability(kind: .processExec, argv: argv, cwd: args["cwd"].str),
            Capability(kind: .network, allowed: network),
        ]
        if !readPaths.isEmpty { capabilities.append(Capability(kind: .fsRead, paths: readPaths)) }
        if !writePaths.isEmpty { capabilities.append(Capability(kind: .fsWrite, paths: writePaths)) }

        let intent = try makeIntent(agent: agent, deviceId: deviceId, goal: args["goal"].str,
                                    request: "run: \(argv.joined(separator: " "))",
                                    capabilities: capabilities)
        let response = try sendIntent(intent, payload: ["wait_ms": .number(Double(waitMs))],
                                      to: link, timeout: TimeInterval(waitMs) / 1000.0 + 630)
        return response
    }

    private func getOutput(_ args: JSONValue) throws -> JSONValue {
        let (_, _, link) = try requireGrantedLink(args)
        guard let handle = args["handle"].str else { throw ToolError("missing 'handle'") }
        // Output retrieval is bound to an already-approved run; no new intent.
        return try link.rpc.call("get_output", [
            "handle": .string(handle),
            "since": .number(Double(args["since"].int ?? 0)),
        ], timeout: 30)
    }

    private func listDeviceTools(_ args: JSONValue) throws -> JSONValue {
        let (_, _, link) = try requireGrantedLink(args)
        return ["tools": link.blessedTools]
    }

    private func useTool(_ args: JSONValue) throws -> JSONValue {
        let (agent, deviceId, link) = try requireGrantedLink(args)
        guard let tool = args["tool"].str else { throw ToolError("missing 'tool'") }
        let intent = try makeIntent(agent: agent, deviceId: deviceId, goal: args["goal"].str,
                                    request: "use blessed tool: \(tool)",
                                    capabilities: [Capability(kind: .tool, tool: tool)])
        let response = try sendIntent(intent, payload: ["args": args["args"]], to: link)
        return ["result": response["result"]]
    }
}

/// Static MCP tool definitions (DESIGN.md §3).
enum MCPTools {
    static let definitions: JSONValue = [
        [
            "name": "list_devices",
            "description": "List the Macs visible to this agent: id, name, online status, and whether you already hold an access grant.",
            "inputSchema": ["type": "object", "properties": [:]],
        ],
        [
            "name": "request_device_access",
            "description": "Ask a Mac's owner for permission to use it. State your goals honestly — the owner sees them. Returns granted or denied.",
            "inputSchema": ["type": "object", "required": ["device", "goals"], "properties": [
                "device": ["type": "string", "description": "Device id from list_devices"],
                "goals": ["type": "string", "description": "What you intend to do on this Mac and why"],
            ]],
        ],
        [
            "name": "read_file",
            "description": "Read a file on a granted Mac. The owner may be asked to approve.",
            "inputSchema": ["type": "object", "required": ["device", "path"], "properties": [
                "device": ["type": "string"],
                "path": ["type": "string", "description": "Absolute path (~ allowed)"],
                "goal": ["type": "string", "description": "Why you need this file (shown to the approver)"],
            ]],
        ],
        [
            "name": "write_file",
            "description": "Write a file on a granted Mac. The owner may be asked to approve.",
            "inputSchema": ["type": "object", "required": ["device", "path", "content"], "properties": [
                "device": ["type": "string"],
                "path": ["type": "string"],
                "content": ["type": "string"],
                "goal": ["type": "string", "description": "Why (shown to the approver)"],
            ]],
        ],
        [
            "name": "run_command",
            "description": "Run a CLI command on a granted Mac inside a sandbox limited to the paths you declare here. Declare every path you need up front; undeclared paths are blocked by the sandbox. Waits up to wait_ms; if still running you get a handle for get_output.",
            "inputSchema": ["type": "object", "required": ["device", "argv"], "properties": [
                "device": ["type": "string"],
                "argv": ["type": "array", "items": ["type": "string"],
                         "description": "Command and arguments, e.g. [\"ls\", \"-la\", \"/tmp\"]"],
                "cwd": ["type": "string", "description": "Working directory (readable by the sandbox)"],
                "read_paths": ["type": "array", "items": ["type": "string"],
                               "description": "Directories/files the command may read"],
                "write_paths": ["type": "array", "items": ["type": "string"],
                                "description": "Directories/files the command may write"],
                "network": ["type": "boolean", "description": "Whether the command needs network access (default false)"],
                "wait_ms": ["type": "integer", "description": "How long to wait for completion before returning a handle (default 10000)"],
                "goal": ["type": "string", "description": "Why (shown to the approver)"],
            ]],
        ],
        [
            "name": "get_output",
            "description": "Fetch incremental output of a still-running command. Pass 'since' = the output_length you last saw.",
            "inputSchema": ["type": "object", "required": ["device", "handle"], "properties": [
                "device": ["type": "string"],
                "handle": ["type": "string"],
                "since": ["type": "integer"],
            ]],
        ],
        [
            "name": "list_device_tools",
            "description": "List the blessed tools available on a specific Mac, with their JSON input schemas. Different Macs have different tools.",
            "inputSchema": ["type": "object", "required": ["device"], "properties": [
                "device": ["type": "string"],
            ]],
        ],
        [
            "name": "use_tool",
            "description": "Invoke a blessed tool on a granted Mac (discover them with list_device_tools).",
            "inputSchema": ["type": "object", "required": ["device", "tool"], "properties": [
                "device": ["type": "string"],
                "tool": ["type": "string"],
                "args": ["type": "object"],
                "goal": ["type": "string", "description": "Why (shown to the approver)"],
            ]],
        ],
    ]
}

File: packages/broker-core/src/mcpSession.ts
/**
 * One authenticated agent connection speaking MCP (JSON-RPC 2.0, one message
 * per line) — twin of DomoBrokerCore/MCPSession.swift. Translates tool calls
 * into signed intents routed to devices.
 */
import crypto from "node:crypto";
import {
  Capability,
  canonicalBytes,
  canonicalJSON,
  Intent,
  JSONValue,
  jv,
  makeIntent,
  signIntent,
} from "@domo/protocol";
import { Connection } from "@domo/transport";
import type { Broker, DeviceLink } from "./broker.js";
import { AgentRecord, agentKeyPair } from "./store.js";

class ToolError extends Error {}

export class MCPSession {
  private readonly sessionId = crypto.randomUUID().toUpperCase();

  constructor(
    private readonly broker: Broker,
    private readonly conn: Connection,
    private readonly agentToken: string,
  ) {}

  private get agent(): AgentRecord | null {
    return this.broker.store.agent(this.agentToken);
  }

  /** The agent identity bound to this session, for revocation targeting. */
  get boundAgentId(): string | null {
    return this.agent?.agentId ?? null;
  }

  closeSession(): void {
    this.conn.close();
  }

  handleLine(line: Buffer): void {
    let message: JSONValue;
    try {
      message = JSON.parse(line.toString("utf8")) as JSONValue;
    } catch {
      return;
    }
    const m = jv(message);
    const id = m.get("id").value ?? null;
    const method = m.get("method").str;
    if (method === null) return;
    if (id === null) return; // notifications (e.g. notifications/initialized)
    void this.dispatch(method, m.get("params").value ?? null).then(
      (result) => this.send({ jsonrpc: "2.0", id, result }),
      (error: unknown) =>
        this.send({
          jsonrpc: "2.0",
          id,
          error: { code: -32000, message: error instanceof Error ? error.message : String(error) },
        }),
    );
  }

  private send(message: JSONValue): void {
    this.conn.sendLine(canonicalBytes(message));
  }

  private async dispatch(method: string, params: JSONValue): Promise<JSONValue> {
    switch (method) {
      case "initialize": {
        const requested = jv(params).get("protocolVersion").str ?? "2024-11-05";
        return {
          protocolVersion: requested,
          capabilities: { tools: {} },
          serverInfo: { name: "domo-broker", version: "0.1.0" },
        };
      }
      case "ping":
        return {};
      case "tools/list":
        return { tools: MCP_TOOLS };
      case "tools/call": {
        const name = jv(params).get("name").str;
        if (name === null) throw new Error("missing tool name");
        return this.callTool(name, jv(params).get("arguments").value ?? null);
      }
      default:
        throw new Error(`method not supported: ${method}`);
    }
  }

  // MARK: Tool dispatch

  private async callTool(name: string, args: JSONValue): Promise<JSONValue> {
    try {
      const result = await this.runTool(name, args);
      return {
        content: [{ type: "text", text: canonicalJSON(result) }],
        isError: false,
      };
    } catch (error: unknown) {
      const message = error instanceof Error ? error.message : String(error);
      return { content: [{ type: "text", text: message }], isError: true };
    }
  }

  private async runTool(name: string, args: JSONValue): Promise<JSONValue> {
    switch (name) {
      case "list_devices":
        return this.listDevices();
      case "request_device_access":
        return this.requestAccess(args);
      case "read_file":
        return this.readFile(args);
      case "write_file":
        return this.writeFile(args);
      case "run_command":
        return this.runCommand(args);
      case "get_output":
        return this.getOutput(args);
      case "list_device_tools":
        return this.listDeviceTools(args);
      case "use_tool":
        return this.useTool(args);
      default:
        throw new ToolError(`unknown tool: ${name}`);
    }
  }

  private requireAgent(): AgentRecord {
    const agent = this.agent;
    if (!agent) throw new ToolError("agent identity revoked");
    if (this.broker.store.isRevoked(agent.agentId)) {
      throw new ToolError("agent access has been revoked");
    }
    return agent;
  }

  private requireGrantedLink(args: JSONValue): {
    agent: AgentRecord;
    deviceId: string;
    link: DeviceLink;
  } {
    const agent = this.requireAgent();
    const deviceId = jv(args).get("device").str;
    if (deviceId === null) throw new ToolError("missing 'device'");
    if (!agent.grantedDevices.includes(deviceId)) {
      throw new ToolError(`no grant for device ${deviceId}; call request_device_access first`);
    }
    const link = this.broker.deviceLink(deviceId);
    if (!link) throw new ToolError(`device ${deviceId} is offline`);
    return { agent, deviceId, link };
  }

  private listDevices(): JSONValue {
    const agent = this.requireAgent();
    const online = this.broker.onlineDeviceIds();
    return {
      devices: this.broker.store.allDevices().map((device) => ({
        id: device.deviceId,
        name: device.name,
        online: online.has(device.deviceId),
        granted: agent.grantedDevices.includes(device.deviceId),
      })),
    };
  }

  private async requestAccess(args: JSONValue): Promise<JSONValue> {
    const agent = this.requireAgent();
    const a = jv(args);
    const deviceId = a.get("device").str;
    if (deviceId === null) throw new ToolError("missing 'device'");
    const goals = a.get("goals").str;
    if (goals === null || goals.length === 0) {
      throw new ToolError("missing 'goals' — state what you intend to do on this Mac");
    }
    const link = this.broker.deviceLink(deviceId);
    if (!link) throw new ToolError(`device ${deviceId} is offline`);
    if (agent.grantedDevices.includes(deviceId)) {
      return { status: "granted", note: "already granted" };
    }
    // Approval can take as long as a human takes; generous timeout.
    const response = await link.rpc.call(
      "access_request",
      {
        agent: {
          id: agent.agentId,
          display: agent.display,
          publicKey: agent.publicKeyBase64,
        },
        goals,
      },
      600,
    );
    const approved = jv(response).get("approved").bool ?? false;
    if (approved) {
      this.broker.store.grantDevice(this.agentToken, deviceId);
      this.broker.store.recordSessionGoals(this.agentToken, goals);
    }
    return { status: approved ? "granted" : "denied" };
  }

  // MARK: Intent construction

  private makeSignedIntent(
    agent: AgentRecord,
    deviceId: string,
    goal: string | undefined,
    request: string,
    capabilities: Capability[],
  ): Intent {
    const intent = makeIntent({
      agentId: agent.agentId,
      agentDisplay: agent.display,
      agentPublicKey: agent.publicKeyBase64,
      deviceId,
      goal,
      planContext: agent.sessionGoals,
      request,
      capabilities,
      sessionId: this.sessionId,
    });
    signIntent(intent, agentKeyPair(agent));
    return intent;
  }

  private async sendIntent(
    intent: Intent,
    payload: JSONValue,
    link: DeviceLink,
    timeoutSeconds = 630,
  ): Promise<JSONValue> {
    const response = await link.rpc.call(
      "intent",
      { intent: intent as unknown as JSONValue, payload },
      timeoutSeconds,
    );
    const status = jv(response).get("status").str;
    if (status === "denied") throw new ToolError("the device owner denied this request");
    if (status === "rejected") {
      throw new ToolError(
        `device rejected intent: ${jv(response).get("reason").str ?? "unknown"}`,
      );
    }
    if (status === "error") {
      throw new ToolError(jv(response).get("error").str ?? "device error");
    }
    return response;
  }

  // MARK: Tools

  private async readFile(args: JSONValue): Promise<JSONValue> {
    const { agent, deviceId, link } = this.requireGrantedLink(args);
    const path = jv(args).get("path").str;
    if (path === null) throw new ToolError("missing 'path'");
    const intent = this.makeSignedIntent(
      agent,
      deviceId,
      jv(args).get("goal").str ?? undefined,
      `read file: ${path}`,
      [{ kind: "fs.read", paths: [path] }],
    );
    const response = await this.sendIntent(intent, null, link);
    const base64 = jv(response).get("content_base64").str;
    if (base64 === null) throw new ToolError("device returned no content");
    const data = Buffer.from(base64, "base64");
    const text = data.toString("utf8");
    // Return text when it round-trips as UTF-8, base64 otherwise (binary safety).
    if (Buffer.from(text, "utf8").equals(data)) {
      return { path, content: text };
    }
    return { path, content_base64: base64 };
  }

  private async writeFile(args: JSONValue): Promise<JSONValue> {
    const { agent, deviceId, link } = this.requireGrantedLink(args);
    const a = jv(args);
    const path = a.get("path").str;
    if (path === null) throw new ToolError("missing 'path'");
    const content = a.get("content").str;
    if (content === null) throw new ToolError("missing 'content'");
    const data = Buffer.from(content, "utf8");
    const intent = this.makeSignedIntent(
      agent,
      deviceId,
      a.get("goal").str ?? undefined,
      `write file: ${path} (${data.length} bytes)`,
      [{ kind: "fs.write", paths: [path] }],
    );
    const response = await this.sendIntent(
      intent,
      { content_base64: data.toString("base64") },
      link,
    );
    return { path, bytes: jv(response).get("bytes").value ?? null };
  }

  private async runCommand(args: JSONValue): Promise<JSONValue> {
    const { agent, deviceId, link } = this.requireGrantedLink(args);
    const a = jv(args);
    const argvValues = a.get("argv").arr;
    if (!argvValues || argvValues.length === 0) throw new ToolError("missing 'argv'");
    const argv = argvValues.filter((v): v is string => typeof v === "string");
    if (argv.length !== argvValues.length) throw new ToolError("argv must be strings");
    const readPaths = (a.get("read_paths").arr ?? []).filter(
      (v): v is string => typeof v === "string",
    );
    const writePaths = (a.get("write_paths").arr ?? []).filter(
      (v): v is string => typeof v === "string",
    );
    const network = a.get("network").bool ?? false;
    const waitMs = a.get("wait_ms").int ?? 10000;

    const capabilities: Capability[] = [
      { kind: "process.exec", argv, cwd: a.get("cwd").str ?? undefined },
      { kind: "network", allowed: network },
    ];
    if (readPaths.length > 0) capabilities.push({ kind: "fs.read", paths: readPaths });
    if (writePaths.length > 0) capabilities.push({ kind: "fs.write", paths: writePaths });

    const intent = this.makeSignedIntent(
      agent,
      deviceId,
      a.get("goal").str ?? undefined,
      `run: ${argv.join(" ")}`,
      capabilities,
    );
    return this.sendIntent(intent, { wait_ms: waitMs }, link, waitMs / 1000 + 630);
  }

  private async getOutput(args: JSONValue): Promise<JSONValue> {
    const { link } = this.requireGrantedLink(args);
    const handle = jv(args).get("handle").str;
    if (handle === null) throw new ToolError("missing 'handle'");
    // Output retrieval is bound to an already-approved run; no new intent.
    return link.rpc.call(
      "get_output",
      { handle, since: jv(args).get("since").int ?? 0 },
      30,
    );
  }

  private listDeviceTools(args: JSONValue): JSONValue {
    const { link } = this.requireGrantedLink(args);
    return { tools: link.blessedTools };
  }

  private async useTool(args: JSONValue): Promise<JSONValue> {
    const { agent, deviceId, link } = this.requireGrantedLink(args);
    const tool = jv(args).get("tool").str;
    if (tool === null) throw new ToolError("missing 'tool'");
    const intent = this.makeSignedIntent(
      agent,
      deviceId,
      jv(args).get("goal").str ?? undefined,
      `use blessed tool: ${tool}`,
      [{ kind: "tool", tool }],
    );
    const response = await this.sendIntent(
      intent,
      { args: jv(args).get("args").value ?? null },
      link,
    );
    return { result: jv(response).get("result").value ?? null };
  }
}

/** Static MCP tool definitions (DESIGN.md §3) — verbatim from the Swift broker. */
export const MCP_TOOLS: JSONValue = [
  {
    name: "list_devices",
    description:
      "List the Macs visible to this agent: id, name, online status, and whether you already hold an access grant.",
    inputSchema: { type: "object", properties: {} },
  },
  {
    name: "request_device_access",
    description:
      "Ask a Mac's owner for permission to use it. State your goals honestly — the owner sees them. Returns granted or denied.",
    inputSchema: {
      type: "object",
      required: ["device", "goals"],
      properties: {
        device: { type: "string", description: "Device id from list_devices" },
        goals: { type: "string", description: "What you intend to do on this Mac and why" },
      },
    },
  },
  {
    name: "read_file",
    description: "Read a file on a granted Mac. The owner may be asked to approve.",
    inputSchema: {
      type: "object",
      required: ["device", "path"],
      properties: {
        device: { type: "string" },
        path: { type: "string", description: "Absolute path (~ allowed)" },
        goal: { type: "string", description: "Why you need this file (shown to the approver)" },
      },
    },
  },
  {
    name: "write_file",
    description: "Write a file on a granted Mac. The owner may be asked to approve.",
    inputSchema: {
      type: "object",
      required: ["device", "path", "content"],
      properties: {
        device: { type: "string" },
        path: { type: "string" },
        content: { type: "string" },
        goal: { type: "string", description: "Why (shown to the approver)" },
      },
    },
  },
  {
    name: "run_command",
    description:
      "Run a CLI command on a granted Mac inside a sandbox limited to the paths you declare here. Declare every path you need up front; undeclared paths are blocked by the sandbox. Waits up to wait_ms; if still running you get a handle for get_output.",
    inputSchema: {
      type: "object",
      required: ["device", "argv"],
      properties: {
        device: { type: "string" },
        argv: {
          type: "array",
          items: { type: "string" },
          description: 'Command and arguments, e.g. ["ls", "-la", "/tmp"]',
        },
        cwd: { type: "string", description: "Working directory (readable by the sandbox)" },
        read_paths: {
          type: "array",
          items: { type: "string" },
          description: "Directories/files the command may read",
        },
        write_paths: {
          type: "array",
          items: { type: "string" },
          description: "Directories/files the command may write",
        },
        network: {
          type: "boolean",
          description: "Whether the command needs network access (default false)",
        },
        wait_ms: {
          type: "integer",
          description: "How long to wait for completion before returning a handle (default 10000)",
        },
        goal: { type: "string", description: "Why (shown to the approver)" },
      },
    },
  },
  {
    name: "get_output",
    description:
      "Fetch incremental output of a still-running command. Pass 'since' = the output_length you last saw.",
    inputSchema: {
      type: "object",
      required: ["device", "handle"],
      properties: {
        device: { type: "string" },
        handle: { type: "string" },
        since: { type: "integer" },
      },
    },
  },
  {
    name: "list_device_tools",
    description:
      "List the blessed tools available on a specific Mac, with their JSON input schemas. Different Macs have different tools.",
    inputSchema: {
      type: "object",
      required: ["device"],
      properties: { device: { type: "string" } },
    },
  },
  {
    name: "use_tool",
    description: "Invoke a blessed tool on a granted Mac (discover them with list_device_tools).",
    inputSchema: {
      type: "object",
      required: ["device", "tool"],
      properties: {
        device: { type: "string" },
        tool: { type: "string" },
        args: { type: "object" },
        goal: { type: "string", description: "Why (shown to the approver)" },
      },
    },
  },
];
